Sources/SourceKitLSP/SourceKitServer.swift

```diff
@@ -1702,73 +1662,112 @@ extension SourceKitServer {
-        occurs = index.occurrences(ofUSR: usr, roles: [.declaration])
+    guard let index = await self.workspaceForDocument(uri: req.textDocument.uri)?.index else {
+      return []
+    }
+    guard let usr = symbol.usr else { return [] }
+    logger.info("performing indexed jump-to-def with usr \(usr)")
+    var occurances = index.occurrences(ofUSR: usr, roles: [.definition])
+    if occurances.isEmpty {
+      occurances = index.occurrences(ofUSR: usr, roles: [.declaration])
+    }
+    if symbol.isDynamic ?? false {
+      lazy var transitiveReceiverUsrs: [String] = transitiveSubtypeClosure(
+        ofUsrs: symbol.receiverUsrs ?? [],
+        index: index
+      )
+      occurances += occurances.flatMap {
+        let overrides = index.occurrences(relatedToUSR: $0.symbol.usr, roles: .overrideOf)
+        // Only contain overrides that are children of one of the receiver types or their subtypes.
+        return overrides.filter { override in
+          override.relations.contains(where: {
+            $0.roles.contains(.childOf) && transitiveReceiverUsrs.contains($0.symbol.usr)
+          })
+        }
       }
-      return occurs
     }
 
-    // if first resolved location is in `.swiftinterface` file. Use moduleName to return
-    // textual interface
-    if let firstResolved = resolved.first,
-      let moduleName = firstResolved.occurrence?.location.moduleName,
-      firstResolved.location.uri.fileURL?.pathExtension == "swiftinterface"
-    {
-      return try await self.definitionInInterface(
-        req,
-        moduleName: moduleName,
-        symbolUSR: firstResolved.occurrence?.symbol.usr,
-        languageService: languageService
-      )
+    return try await occurances.asyncCompactMap { occurance in
+      if URL(fileURLWithPath: occurance.location.path).pathExtension == "swiftinterface" {
+        // if first resolved location is in `.swiftinterface` file. Use moduleName to return
+        // textual interface
+        return try await self.definitionInInterface(
+          req,
+          moduleName: occurance.location.moduleName,
+          symbolUSR: occurance.symbol.usr,
+          languageService: languageService
+        )
+      }
+      return indexToLSPLocation(occurance.location)
     }
-    let locs = resolved.map(\.location)
+  }
+
+  func definition(
+    _ req: DefinitionRequest,
+    workspace: Workspace,
+    languageService: ToolchainLanguageServer
+  ) async throws -> LocationsOrLocationLinksResponse? {
+    let indexBasedResponse = try await indexBasedDefinition(req, workspace: workspace, languageService: languageService)
     // If we're unable to handle the definition request using our index, see if the
     // language service can handle it (e.g. clangd can provide AST based definitions).
-    if locs.isEmpty {
-      return try await languageService.definition(req)
+    if indexBasedResponse.isEmpty {
+      do {
+        return try await languageService.definition(req)
+      } catch {}
```

Comment:
Why are we ignoring the error here?

Reply:
Changed to `orLog` and added the following comment:

> We are on only calling the language service's `definition` function if your index-based lookup failed.  If this fallback request fails, its error is usually not very enlightening. For example the `SwiftLanguageServer` will always respond with `unsupported method`. Thus, only log such a failure instead of returning it to the client.